tensorflow_lite_support/python/task/audio/core/tensor_audio.py

```diff
@@ -0,0 +1,132 @@
+# limitations under the License.
+"""TensorAudio class."""
+
+import numpy as np
+
+from tensorflow_lite_support.python.task.audio.core.pybinds import _pywrap_audio_buffer
+from tensorflow_lite_support.python.task.audio.core import audio_record
+
+_CppAudioFormat = _pywrap_audio_buffer.AudioFormat
+_LoadAudioBufferFromFile = _pywrap_audio_buffer.LoadAudioBufferFromFile
+
+
+class TensorAudio(object):
+  """A wrapper class to store the input audio."""
+
+  def __init__(self,
+               audio_format: _CppAudioFormat,
+               buffer_size: int) -> None:
+    """Initializes the `TensorAudio` object.
+
+    Args:
+      audio_format: format of the audio.
+      buffer_size: buffer size of the audio.
+    """
+    self._format = audio_format
+    self._buffer_size = buffer_size
+    self._buffer = np.zeros(
+        [self._buffer_size, self._format.channels], dtype=np.float32)
+
+  def clear(self):
+    """Clear the internal buffer and fill it with zeros."""
+    self._buffer.fill(0)
+
+  @classmethod
+  def create_from_wav_file(cls,
+                           file_name: str,
+                           buffer_size: int) -> "TensorAudio":
+    """Creates `TensorAudio` object from the WAV file.
+
+    Args:
+      file_name: WAV file name.
+      buffer_size: Required input buffer size.
+
+    Returns:
+      `TensorAudio` object.
+
+    Raises:
+      status.StatusNotOk if the audio file can't be decoded.
+    """
+    # TODO(b/220931229): Raise RuntimeError instead of status.StatusNotOk.
+    # Need to import the module to catch this error:
+    # `from pybind11_abseil import status`
+    # see https://github.com/pybind/pybind11_abseil#abslstatusor.
+    audio = _LoadAudioBufferFromFile(
+      file_name, buffer_size, np.zeros([buffer_size]))
+    tensor = TensorAudio(audio.audio_format, audio.buffer_size)
+    tensor.load_from_array(np.array(audio.float_buffer, copy=False))
+    return tensor
+
+  def load_from_audio_record(self, record: audio_record.AudioRecord) -> None:
+    """Loads audio data from an AudioRecord instance.
+    Args:
+      record: An AudioRecord instance.
+    Raises:
+      ValueError: Raised if the audio record's config is invalid.
+    """
+    if record.buffer_size < self._buffer_size:
+      raise ValueError(
+        "The audio record's buffer size cannot be smaller than the tensor "
+        "audio's sample count.")
+
+    if record.channels != self._format.channels:
+      raise ValueError(
+        f"The audio record's channel count doesn't match. "
+        f"Expects {self._format.channels} channel(s).")
+
+    if record.sampling_rate != self._format.sample_rate:
+      raise ValueError(
+        f"The audio record's sampling rate doesn't match. "
+        f"Expects {self._format.sample_rate}Hz.")
+
+    # Load audio data from the AudioRecord instance.
+    data = record.read(self._buffer_size)
+    self.load_from_array(data.astype(np.float32))
+
+  def load_from_array(self, src: np.ndarray) -> None:
```

Comment:
ditto

Reply:
@khanhlvg I've ported the Java implementation to Python. It should look something like this, right?

```python
def load_from_array(self,
                      src: np.ndarray,
                      offset: int = None,
                      size: int = None) -> None:
    """Loads the entire float NumPy array or a slice to the ring buffer."""
    if len(src) != len(self._buffer):
      raise ValueError(
        f"Input audio contains an invalid number of samples. "
        f"Expect {len(self._buffer)}.")
    elif src.shape[1] != self._format.channels:
      raise ValueError(
        f"Input audio contains an invalid number of channels. "
        f"Expect {self._format.channels}.")

    if offset is None and size is None:
      # Loads the entire float array to the ring buffer. If the float array is
      # longer than ring buffer's capacity, samples with lower indices in the
      # array will be ignored.
      self._buffer = src
    else:
      # Loads a slice of the float array to the ring buffer. If the float array
      # is longer than ring buffer's capacity, samples with lower indices in
      # the array will be ignored.
      buffer_size = len(self._buffer)
      next_index = self._next_index

      if size > buffer_size:
        offset = size - buffer_size
        size = buffer_size

      if next_index + size < buffer_size:
        # No need to wrap `next_index`.
        self._buffer[next_index:next_index + size] = src[offset:offset + size]
      else:
        # Need to wrap `next_index`, perform copy in two chunks.
        first_chunk_size = buffer_size - next_index
        self._buffer[next_index:buffer_size] = \
          src[offset:offset + first_chunk_size]
        self._buffer[0:size - first_chunk_size] = \
          src[offset + first_chunk_size:offset + size]

      self._next_index = (next_index + size) % buffer_size
```

Reply:
@kinaryml 
* The TensorAudio buffer size can be different to the src array, hence you shouldn't raise a ValueError when they're different.
* You should check if offset + size >= len(src) and raise a ValueError if it is.
* The size of the TensorAudio buffer shouldn't change after loading an array. Therefore, you can't do `self._buffer = src` but you have to check:
   * size > len(self._buffer): loading only the buffer size worth of samples from the end of the src array (`new_offset = offset + size - len(self._buffer)` and `new_size = len(self._buffer)`).
   * size < len(self._buffer): load the array into the end of the buffer. 
* Regarding the implementation of the buffer, you can use np.roll to make the code easier to read. There's no need to implement the RingBuffer exactly like the Java (storing the buffer begin index as next_index) because Java doesn't have convenient method like np.roll.

Reply:
@khanhlvg Thanks for the feedback. Yeah the `offset + size > buffer_size` was a typo. Fixed it now. I'll take a look at the rest.